verilog/analysis/checkers/disable_fork_rule_test.cc

```diff
@@ -0,0 +1,99 @@
+// Copyright 2017-2020 The Verible Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "verilog/analysis/checkers/disable_fork_rule.h"
+
+#include <initializer_list>
+
+#include "common/analysis/linter_test_utils.h"
+#include "common/analysis/syntax_tree_linter_test_utils.h"
+#include "common/text/symbol.h"
+#include "gtest/gtest.h"
+#include "verilog/CST/verilog_nonterminals.h"
+#include "verilog/CST/verilog_treebuilder_utils.h"
+#include "verilog/analysis/verilog_analyzer.h"
+#include "verilog/parser/verilog_token_enum.h"
+
+namespace verilog {
+namespace analysis {
+namespace {
+
+using verible::LintTestCase;
+using verible::RunLintTestCases;
+
+TEST(DisableStatementTest, FunctionFailures) {
+  constexpr int kToken = TK_disable;
+  const std::initializer_list<LintTestCase> kDisableStatementTestCases = {
+      {""},
+      {"module m;\nendmodule\n"},
+      {"module m;\ninitial begin end\nendmodule"},
+      {"module m;\ninitial begin fork\n", "begin\n#6;\nend\n",
+       "begin\n#3;\nend\n", "join_any\n", "disable fork;\n", "end\nendmodule"},
+      {"module m;\ninitial begin fork\n", "begin : foo\n", "disable foo;\n",
+       "end\n", "join_any\n", "end\nendmodule"},
+      {"module m;\ninitial begin\n fork\n", "begin : foo\n", "begin : foo_2\n",
+       "disable foo_2;\n", "end\n", "end\n", "join_any\n", "end\nendmodule"},
+      {"module m;\ninitial begin\n fork\n", "begin : foo\n", "begin : foo_2\n",
+       "disable foo;\n", "end\n", "end\n", "join_any\n", "end\nendmodule"},
+      {"module m;\ninitial begin fork\n",
+       "begin\n#6;\nend\n",
+       "begin\n#3;\nend\n",
+       "join_any\n",
+       {kToken, "disable"},
+       " fork_invalid;\n",
+       "end\nendmodule"},
+      {"module m;\n",
+       "initial begin:foo;\n",
+       "fork\n",
+       "begin : fo\n",
+       {kToken, "disable"},
```

Comment:
this should actually be allowed. `disable foo` refers to the begin block above, not a labelled fork.

Reply:
But this is not a sequential block since this is a `initial begin` block?
I make such assumption based on the clarification written here https://github.com/google/verible/issues/585#issuecomment-769616018
And in the last paragraph sriyerg wrote:
```
Also, I would assume that attempting to disable process execution threads such as initial, always* and final using their labels:

  initial begin : test
    ...
    if (cond) disable test;
  end

is illegal because these are not considered sequential code-blocks. The LRM makes no mention of validity of this (at least I couldn't find it) . So these will need to be flagged as well.
```
So should they be tried as illegal as well?